demos/python_demos/asl_recognition_demo/asl_recognition_demo/video_stream.py

```diff
@@ -0,0 +1,154 @@
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import time
+import signal
+from os.path import exists
+from multiprocessing import Process, Array
+
+import cv2
+import numpy as np
+
+
+class VideoStream:
+    """ This class returns constant framerate frames from the input stream. """
+
+    def __init__(self, path, trg_fps, batch_size):
+        """Conctructor"""
+
+        if not exists(path):
+            raise Exception('File not found: {}'.format(path))
+
+        self._trg_fps = trg_fps
+        assert self._trg_fps > 0
+        self._batch_size = batch_size
+        assert self._batch_size > 0
+
+        self._cap = cv2.VideoCapture(path)
+        assert self._cap.isOpened(), "Can't open " + path
+        self._cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)
+
+        self._image_shape = [int(self._cap.get(cv2.CAP_PROP_FRAME_HEIGHT)),
+                             int(self._cap.get(cv2.CAP_PROP_FRAME_WIDTH)),
+                             3]
+        self._batch_shape = [batch_size] + self._image_shape
+
+        self._image_buffer_size = int(np.prod(self._image_shape))
+        self._batch_buffer_size = int(np.prod(self._batch_shape))
+
+        self._raw_frame = Array('B', self._image_buffer_size, lock=True)
+        self._slow_frame = Array('B', self._image_buffer_size, lock=True)
+        self._slow_batch = Array('B', self._batch_buffer_size, lock=True)
+
+    def get_live_frame(self):
+        """Returns last live frame from the input stream"""
+
+        with self._raw_frame.get_lock():
+            buffer = np.frombuffer(self._raw_frame.get_obj(), dtype=np.uint8)
+            frame = np.copy(buffer.reshape(self._image_shape))
+        return frame
+
+    def get_frame(self):
+        """Returns last frame with constant framerate from the input stream"""
+
+        with self._slow_frame.get_lock():
+            buffer = np.frombuffer(self._slow_frame.get_obj(), dtype=np.uint8)
+            frame = np.copy(buffer.reshape(self._image_shape))
+        return frame
+
+    def get_batch(self):
+        """Returns last batch of frames with constant framerate from the input stream"""
+
+        with self._slow_batch.get_lock():
+            buffer = np.frombuffer(self._slow_batch.get_obj(), dtype=np.uint8)
+            batch = np.copy(buffer.reshape(self._batch_shape))
+        return batch
+
+    def start(self):
+        """Starts internal threads"""
+
+        self._frame_generator_process = \
+            Process(target=self._frame_generator,
+                    args=(self._cap, self._raw_frame, self._image_shape))
```

Comment:
Does this work on Windows? I seriously doubt that `self._cap` is picklable.

Reply:
Fixed.